File: sdk/go/internal/tx/chain_adapter.go
package tx

import (
	"context"
	"fmt"

	"github.com/seocheon/sdk-go/internal/chain"
)

// ChainClientAdapter adapts chain.Client to the ChainQuerier interface.
type ChainClientAdapter struct {
	client chain.Client
}

// NewChainClientAdapter creates a new adapter wrapping a chain.Client.
func NewChainClientAdapter(client chain.Client) *ChainClientAdapter {
	return &ChainClientAdapter{client: client}
}

func (a *ChainClientAdapter) GetAccountInfo(ctx context.Context, address string) (uint64, uint64, error) {
	info, err := a.client.GetAccountInfo(ctx, address)
	if err != nil {
		return 0, 0, fmt.Errorf("querying account info: %w", err)
	}
	return info.AccountNumber, info.Sequence, nil
}

func (a *ChainClientAdapter) BroadcastTxSync(ctx context.Context, txBytes []byte) (string, uint32, string, error) {
	// The chain client expects raw bytes; it handles base64 encoding internally
	resp, err := a.client.BroadcastTx(ctx, txBytes, "sync")
	if err != nil {
		return "", 0, "", fmt.Errorf("broadcasting tx: %w", err)
	}
	return resp.TxHash, resp.Code, resp.RawLog, nil
}

func (a *ChainClientAdapter) GetTxResult(ctx context.Context, txHash string) (*TxResult, error) {
	resp, err := a.client.GetTx(ctx, txHash)
	if err != nil {
		return nil, err
	}

	events := make([]TxEvent, len(resp.Events))
	for i, e := range resp.Events {
		attrs := make([]EventAttribute, len(e.Attributes))
		for j, attr := range e.Attributes {
			attrs[j] = EventAttribute{Key: attr.Key, Value: attr.Value}
		}
		events[i] = TxEvent{Type: e.Type, Attributes: attrs}
	}

	return &TxResult{
		TxHash:    resp.TxHash,
		Height:    resp.Height,
		Code:      resp.Code,
		GasUsed:   resp.GasUsed,
		GasWanted: resp.GasWanted,
		RawLog:    resp.RawLog,
		Events:    events,
	}, nil
}

File: sdk/python/seocheon/internal/tx/chain_adapter.py
"""ChainClient to ChainQuerier adapter."""

from __future__ import annotations

from typing import TYPE_CHECKING

from seocheon.internal.tx.types import TxEventAttr, TxEventData, TxResult

if TYPE_CHECKING:
    from seocheon.infrastructure.chain_client import ChainClient


class ChainClientAdapter:
    """Adapts ChainClient to the ChainQuerier interface."""

    def __init__(self, client: ChainClient) -> None:
        self._client = client

    async def get_account_info(self, address: str) -> tuple[int, int]:
        info = await self._client.get_account_info(address)
        return info["account_number"], info["sequence"]

    async def broadcast_tx_sync(self, tx_bytes: bytes) -> tuple[str, int, str]:
        resp = await self._client.broadcast_tx(tx_bytes, "sync")
        return resp["tx_hash"], resp["code"], resp["raw_log"]

    async def get_tx_result(self, tx_hash: str) -> TxResult:
        resp = await self._client.get_tx(tx_hash)
        events = [
            TxEventData(
                type=e["type"],
                attributes=[TxEventAttr(key=a["key"], value=a["value"]) for a in e.get("attributes", [])],
            )
            for e in resp.get("events", [])
        ]
        return TxResult(
            tx_hash=resp.get("tx_hash", ""),
            height=resp.get("height", 0),
            code=resp.get("code", 0),
            gas_used=resp.get("gas_used", 0),
            gas_wanted=resp.get("gas_wanted", 0),
            raw_log=resp.get("raw_log", ""),
            events=events,
        )
